src/app/cras/edit-cra/invoice-form/invoice-form.component.ts

```diff
@@ -0,0 +1,145 @@
+    new InputConfig(
+      'invoiceConsultantVATToggle',
+      {
+        parent: 'consultant',
+        binding: 'vatExemption',
+        toggle: 'invoiceConsultantVATNumberInput'
+      }
+    ),
+
+
+    new InputConfig('invoiceClientNameInput', { parent: 'client', binding: 'name'}),
+    new InputConfig('invoiceClientAddressInput', { parent: 'client', binding: 'address'}),
+    new InputConfig('invoiceClientTelephoneInput', { parent: 'client', binding: 'telephone'}),
+    new InputConfig('invoiceClientSIRENInput', { parent: 'client', binding: 'siren'}),
+    new InputConfig(
+      'invoiceClientTradeAndCompaniesRegisterCityInput',
+      {
+        parent: 'client',
+        binding: 'tradeAndCompaniesRegisterCity'
+      }
+    ),
+    new InputConfig(
+      'invoiceClientTradeAndCompaniesRegisterToggle',
+      {
+        parent: 'client',
+        binding: 'tradeAndCompaniesRegisterExemption',
+        toggle: 'invoiceClientTradeAndCompaniesRegisterCityInput'
+      }
+    ),
+    new InputConfig('invoiceClientVATNumberInput', { parent: 'client', binding: 'vatNumber' }),
+    new InputConfig(
+      'invoiceClientVATToggle',
+      {
+        parent: 'client',
+        binding: 'vatExemption',
+        toggle: 'invoiceClientVATNumberInput'
+      }
+    ),
+
+
+    new InputConfig('invoicePaymentDateInput', { binding: 'paymentDate'}),
+    new InputConfig('invoicePaymentModalityInput', { binding: 'paymentModality'}),
+    new InputConfig('invoicePaymentLatePenaltyToggle', { binding: 'invoicePaymentLatePenalty'}),
+    new InputConfig('invoiceBankDetailsIBANInput', { binding: 'bankIBAN'}),
+    new InputConfig('invoiceBankDetailsSWIFTInput', { binding: 'bankSWIFT'}),
+    new InputConfig('invoiceBankDetailsDomiciliationInput', { binding: 'bankingDomiciliation'}),
+  ];
+
+  constructor(
+    private formBuilder: FormBuilder,
+    private invoiceDataService: InvoiceDataService
+  ) {}
+
+
+  ngOnInit() {
+    this.invoice =  this.invoiceDataService.getInvoice();
+    this.invoiceForm = this.registerFormControls();
+    this.intAutoFormUpdater();
+  }
+
+  registerFormControls(): FormGroup {
+    const controls: FormControl[] = [];
+
+    this.controlsBindings.forEach(inputConfiguration => {
+      controls[inputConfiguration.inputName] = new FormControl(inputConfiguration.resolveBinding(this.invoice));
+
+
+      Object.defineProperty(this, inputConfiguration.inputName, {
+        get: () => this.invoiceForm.get(inputConfiguration.inputName),
+      });
+
+      controls[inputConfiguration.inputName].valueChanges.subscribe((value: any) => {
+        inputConfiguration.updateBinded(this.invoice, value);
+
+        if (inputConfiguration.config.hasOwnProperty('toggle')) {
+          this.toggleControl(this[inputConfiguration.config.toggle], value);
+        }
+
+        this.invoiceDataService.setInvoice(this.invoice);
+      });
+    });
+
+    return this.formBuilder.group(controls);
+  }
+
+  intAutoFormUpdater(): void {
```

Comment:
I dont understand this method

Reply:
This method subscribes to the observable that changes each time another component changes the invoice. Then update the form.

Reply:
Having to manually update things when data changes seems wrong.